app/Providers/ValidationServiceProvider.php

```diff
@@ -46,44 +46,56 @@ public function boot()
+
+        /**
+         * Unique only if undeleted.
+         *
+         * This works around the use case where multiple deleted items have the same unique attribute.
+         * (I think this is a bug in Laravel's validator?)
+         *
+         * $attribute is the FIELDNAME you're checking against
+         * $value is the VALUE of the item you're checking against the existing values in the fieldname
+         * $parameters[0] is the TABLE NAME you're querying
+         * $parameters[1] is the ID of the item you're querying - this makes it work on saving, checkout, etc,
+         *   since it defaults to 0 if there is no item created yet (new item), but populates the ID if editing
+         *
+         * The UniqueUndeletedTrait prefills these parameters, so you can just use
+         * `unique_undeleted:table,fieldname` in your rules out of the box
+         */
         Validator::extend('unique_undeleted', function ($attribute, $value, $parameters, $validator) {
+
             if (count($parameters)) {
-                $count = DB::table($parameters[0])->select('id')->where($attribute, '=', $value)->whereNull('deleted_at')->where('id', '!=', $parameters[1])->count();
+
+                // This is a bit of a shim, but serial doesn't have any other rules around it other than that it's nullable
+                if (($parameters[0]=='assets') && ($attribute == 'serial') && (Setting::getSettings()->unique_serial != '1')) {
+                    return true;
+                }
+
+                $count = DB::table($parameters[0])
+                    ->select('id')
+                    ->where($attribute, '=', $value)
+                    ->whereNull('deleted_at')
+                    ->where('id', '!=', $parameters[1])->count();
 
                 return $count < 1;
             }
         });
 
         // Unique if undeleted for two columns
-            // Same as unique_undeleted but taking the combination of two columns as unique constrain.
-            Validator::extend('two_column_unique_undeleted', function ($attribute, $value, $parameters, $validator) {
-                if (count($parameters)) {
-                    $count = DB::table($parameters[0])
-                             ->select('id')->where($attribute, '=', $value)
-                             ->whereNull('deleted_at')
-                             ->where('id', '!=', $parameters[1])
-                             ->where($parameters[2], $parameters[3])->count();
-
-                    return $count < 1;
-                }
-            });
-
-
-        Validator::extend('unique_serial', function ($attribute, $value, $parameters, $validator) {
-            if(Setting::getSettings()->unique_serial == '1') {
-                $count = DB::table('assets')->select('id')->where('serial', '=', $value)->whereNull('deleted_at')->count();
```

Comment:
And, as I read it, I think *this* is the 'smoking gun' for the actual bug in question - there's no test here for whether or not the thing you are currently editing actually *is* the source of the "serial number conflict." (i.e. the check for `id != $parameters[1]`). When used in a FormRequest context, you wouldn't necessarily _have_ an ID at hand, so I can see why we would make that choice - but it does mess us up when we're doing validation at the Model level.